### ISSUE TITLE
Implement a versioning scheme for JITaaS

### ISSUE THREAD

**Issue by mpirvu** [eclipse/openj9#4467]
The JITaaS client should not be allowed to connect to an incompatible JITaaS server, thus we need to keep some version information: major.minor.patch.
For compatibility, the 'major' and 'minor' number should be the same, while the 'patch' number can be different.
Every time when we deliver some changes that break the compatibility, the 'minor' number should be incremented.

**Comment by fjeremic** [comment]
So effectively we'll be following semantic versioning [1]?

[1] https://semver.org/

**Comment by mpirvu** [comment]
@satbir-ca As discussed, please post your proposed design here before starting to modify the code.

**Comment by harryyu1994** [comment]
Looking into this. Will post a summary of the design here.

**Comment by harryyu1994** [comment]
Will likely migrate things from https://github.com/eclipse/openj9/pull/5426/files and create a brand new PR.

**Comment by harryyu1994** [comment]
Proposed Design:

Client sends the version information after it establishes connection with the Server.
There will be at the maximum 7 connections in total for any given client.

Server sends back a message which will inform the client whether they are compatible or not. Client will switch to local compilations if incompatible.

A question, will there ever be a case where a compatible server come up later? I think there are cases where the server that the client talks to could change dynamically while the client is running. For example, when you have a kubernetes endpoint instead of a static ip address for the server. So i think it would be better to have the client check with the server periodically.

**Comment by harryyu1994** [comment]
Design for server side message:
Currently what we have:
```
stream->write(JITaaS::J9ServerMessageType::CHTable_getClassInfoUpdates, JITaaS::Void());
auto recv = stream->read<std::string, std::string>();
const std::string &chtableUnloads = std::get<0>(recv);
const std::string &chtableMods = std::get<1>(recv);
```
A `stream->write` is always followed by a `stream->read` and it always expects to receive the right message.
We are going to associate every response that server gets with a message type like this:
```
stream->write(JITaaS::J9ServerMessageType::CHTable_getClassInfoUpdates, JITaaS::Void());
auto recv = readClientMessage(JITaaS::J9ServerMessageType::CHTable_getClassInfoUpdates);
```
- `readClientMessage` will check if the received message is of the right type.
- we can probably move the `stream->write` part into the `readClientMessage` method as well so it becomes:
```
auto recv = queryClient(JITaaS::J9ServerMessageType::CHTable_getClass, JITaaS::Void());
```

**Comment by mpirvu** [comment]
You can eliminate `uint32 version = 3;` from the definition of the client message. The version could be part of the `AnyData data = 2;` field.

Also, it occurred to me that the version may need to include configuration parameters like, compressed vs non-compressed or JCL level (Java 8, 9, 10, 11, etc). This is because we don't want a client built on Java8 to use a server built on Java11 or viceversa.
Thus, we may need to extent the version info to 64 bits.

**Comment by harryyu1994** [comment]
If we put the version inside `AnyData data = 2`, we would definitely want it to be the first element of the tuple right? that means when it's a `JITaaS::J9ServerMessageType::compilationRequestWithVersion` instead of `JITaaS::J9ServerMessageType::compilationRequest`, we have to shift all the elements in the tuple by one. i.e `clientId = std::get<1>(req);  uint32_t romMethodOffset  = std::get<2>(req);`

```
auto req = stream->read<uint64_t, uint32_t, J9Method *, J9Class*, TR_Hotness, std::string,
         J9::IlGeneratorMethodDetailsType, std::vector<TR_OpaqueClassBlock*>,
         JITaaSHelpers::ClassInfoTuple, std::string, std::string, uint32_t, bool>();

      clientId                  = std::get<0>(req);
      uint32_t romMethodOffset  = std::get<1>(req);
      J9Method *ramMethod       = std::get<2>(req);
      J9Class *clazz            = std::get<3>(req);
      TR_Hotness optLevel       = std::get<4>(req);
      std::string detailsStr    = std::get<5>(req);
      auto detailsType          = std::get<6>(req);
      auto &unloadedClasses     = std::get<7>(req);
      auto &classInfoTuple      = std::get<8>(req);
      std::string clientOptStr  = std::get<9>(req);
      std::string recompInfoStr = std::get<10>(req);
      uint32_t seqNo            = std::get<11>(req); // sequence number at the client
      useAotCompilation         = std::get<12>(req);
```

**Comment by harryyu1994** [comment]
I noticed in Satbir's code he has a version field, and it seems like `clear_version` will make the version data to not be sent. If this is the case then I think having a version field is a better idea.
```
if (_versionCheckStatus == NOTDONE)
         {
         _cMsg.set_version(MAJOR_NUMBER << 24 | MINOR_NUMBER << 8 | PATCH_NUMBER);
         }
      else
         {
         // Need to clear version so version data is not sent.
         _cMsg.clear_version();
         }
```

**Comment by harryyu1994** [comment]
For the same message:
empty version field = 2410
without version field = 2410
with version field set = 2415

It does not affect message size when it's not set.

**Comment by mpirvu** [comment]
Closing this issue as functionality has been delivered.
